ion-schema/src/main/kotlin/com/amazon/ionschema/model/Isl.kt

```diff
@@ -0,0 +1,210 @@
+fun anyOf(vararg types: AstType) = AnyOfConstraint(types.toList())
+fun anyOf(types: Iterable<AstType>) = AnyOfConstraint(types)
+fun byteLength(range: DiscreteRange) = ByteLengthConstraint(range)
+fun byteLength(value: Int) = ByteLengthConstraint(DiscreteRange(value, value))
+fun codepointLength(range: DiscreteRange) = CodepointLengthConstraint(range)
+fun codepointLength(value: Int) = CodepointLengthConstraint(DiscreteRange(value, value))
+fun containerLength(range: DiscreteRange) = ContainerLengthConstraint(range)
+fun containerLength(value: Int) = ContainerLengthConstraint(DiscreteRange(value, value))
+fun contains(vararg values: IonElement) = ContainsConstraint(values.toList())
+fun contains(values: Collection<IonElement>) = ContainsConstraint(values)
+fun element(type: AstType) = ElementConstraint(type)
+fun fields(vararg fields: Pair<String, AstVariablyOccurringType>) = FieldsConstraint(fields.toMap())
+fun fields(fields: Map<String, AstVariablyOccurringType>) = FieldsConstraint(fields)
+fun not(type: AstType) = NotConstraint(type)
+fun oneOf(vararg types: AstType) = OneOfConstraint(types.toList())
+fun oneOf(types: Iterable<AstType>) = OneOfConstraint(types)
+fun orderedElements(vararg types: AstVariablyOccurringType) = OrderedElementsConstraint(types.toList())
+fun orderedElements(types: Iterable<AstVariablyOccurringType>) = OrderedElementsConstraint(types)
+fun precision(range: DiscreteRange) = PrecisionConstraint(range)
+fun precision(value: Int) = PrecisionConstraint(DiscreteRange(value, value))
+fun regex(pattern: String, options: Set<RegexConstraint.Options>) = RegexConstraint(pattern, options)
+fun regex(pattern: String, vararg options: RegexConstraint.Options) = RegexConstraint(pattern, options.toSet())
+fun scale(range: DiscreteRange) = ScaleConstraint(range)
+fun scale(value: Int) = ScaleConstraint(DiscreteRange(value, value))
+fun timestampPrecision(range: TimestampPrecisionRange) = TimestampPrecisionConstraint(range)
+fun timestampPrecision(value: TimestampPrecisions) = TimestampPrecisionConstraint(TimestampPrecisionRange(value, value))
+fun timestampOffset(offset: TimestampOffsetConstraint.Offset) = TimestampOffsetConstraint(offset)
+fun timestampOffset(hours: Int, minutes: Int = 0) = TimestampOffsetConstraint(TimestampOffsetConstraint.Offset.Minutes(hours * 60 + minutes))
+fun timestampOffset(minutes: Int) = TimestampOffsetConstraint(TimestampOffsetConstraint.Offset.Minutes(minutes))
+fun type(type: AstType) = TypeConstraint(type)
+fun utf8ByteLength(range: DiscreteRange) = Utf8ByteLengthConstraint(range)
+fun utf8ByteLength(value: Int) = Utf8ByteLengthConstraint(DiscreteRange(value, value))
+fun validValues(values: Iterable<ValidValue>) = ValidValuesConstraint(values)
+fun validValues(vararg values: ValidValue) = ValidValuesConstraint(values.toList())
+fun validValues(vararg ionValues: IonElement) = ValidValuesConstraint(ionValues.map { ValidValue.SingleValue(it) })
```

Comment:
This should also include functions to construct `valid_values` from a range (timestamp/number)?

Reply:
`fun validValues(vararg values: ValidValue)` will get you pretty close already.

```kotlin
validValues(ValidValue.NumberRange(/* ... */))
validValues(ValidValue.TimestampRange(/* ... */))
```
Or, if we added convenience functions for creating these ranges:
```kotlin
validValues(range(BigDecimal.ONE, Max))
```

That being said, this file (`Isl.kt`) is meant as an example to show that a DSL is possible, but I think it would be better to copy what the AWS Kotlin SDK does and provide Java-style builders _and_ Kotlin-style DSL.